File: leetcode/1122_relative_sort_array.cpp
// Copyright (c) 2019 Jacopo Notarstefano

#include <algorithm>
#include <cassert>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
 public:
    vector<int> relativeSortArray(vector<int>& arr1, vector<int>& arr2) {
        vector<int> result(arr1.begin(), arr1.end());

        unordered_map<int, int> ranks = toRanks(arr2);
        sort(result.begin(), result.end(), [&ranks](int x, int y) {
            if (ranks.find(x) != ranks.end() && ranks.find(y) != ranks.end()) {
                return ranks.at(x) < ranks.at(y);
            } else if (ranks.find(x) != ranks.end()) {
                return true;
            } else if (ranks.find(y) != ranks.end()) {
                return false;
            } else {
                return x < y;
            }
        });

        return result;
    }

 private:
    unordered_map<int, int> toRanks(vector<int>& arr) {
        unordered_map<int, int> result;

        for (int i = 0; i < arr.size(); i++) {
            result[arr[i]] = i;
        }

        return result;
    }
};

int main() {
    auto solution = Solution();

    vector<int> arr1 = {2, 3, 1, 3, 2, 4, 6, 7, 9, 2, 19};
    vector<int> arr2 = {2, 1, 4, 3, 9, 6};

    vector<int> expected = {2, 2, 2, 1, 4, 3, 3, 9, 6, 7, 19};
    vector<int> result = solution.relativeSortArray(arr1, arr2);

    assert(expected == result);
}

File: leetcode/1122_relative_sort_array.py
# -*- coding: utf-8 -*-


class Solution:
    def relativeSortArray(self, arr1, arr2):
        ranks = {el: i for i, el in enumerate(arr2)}
        return sorted(arr1, key=lambda el: (ranks.get(el, float("inf")), el))


if __name__ == "__main__":
    solution = Solution()

    assert [2, 2, 2, 1, 4, 3, 3, 9, 6, 7, 19] == solution.relativeSortArray(
        [2, 3, 1, 3, 2, 4, 6, 19, 9, 2, 7], [2, 1, 4, 3, 9, 6]
    )
